### PR TITLE
[Fix-16680][ApiServer] Query project list paging error

### LINKED ISSUE #16680: [Bug] [ApiServer] Query project list paging error
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

image

### What you expected to happen

The reason is that projectCodes was not checked for null values.

### How to reproduce

When the input query condition does not return any project.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by shouwangyw**
## Purpose of the pull request
This pull request  fix the project query error
Fix #16680

## Brief change log

Change WorkflowDefinitionMapper.xml#queryProjectWorkflowDefinitionCountByProjectCodes

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml [hunk 1]
```diff
@@ -200,10 +200,14 @@
             project_code as projectCode
             ,count(*) as count
         from t_ds_workflow_definition
-        where project_code in
-        <foreach collection="projectCodes" index="index" item="i" open="(" separator="," close=")">
-            #{i}
-        </foreach>
+        <where>
```
Comment (SbloodyS) [inline dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: We should not do this query while projectCodes is null

**Review by SbloodyS — CHANGES_REQUESTED** [review]

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml (hunk 1 above)
Comment (shouwangyw) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: > We should not do this query while projectCodes is null

Is it not possible to query all project when projectCodes is null？？

Comment (SbloodyS) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: This is used to query the number of workflow definitions under each project. If the projects don't exist, the number of workflow definitions under the projects is meaningless.

Comment (ruanwenjun) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: Yes, we should filter this in upper layer. The project_code should avoid to be null.

Comment (shouwangyw) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: > Yes, we should filter this in upper layer. The project_code should avoid to be null.

Yes, but this method must also ensure robustness.

Comment (SbloodyS) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: We don't need this since if the the projects don't exist, the number of workflow definitions under the projects is meaningless.

Comment (shouwangyw) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: > We don't need this since if the the projects don't exist, the number of workflow definitions under the projects is meaningless.

OK, i have fix it use choose-when-otherwise, maybe this is better.

Comment (SbloodyS) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: I think we don't need to change this since we've already filtered it at the entrance.

Comment (shouwangyw) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: > I think we don't need to change this since we've already filtered it at the entrance.

However, if someone uses this method in the future, there will be error.

Comment (SbloodyS) [reply dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: > However, if someone uses this method in the future, there will be error.

We shouldn't do double check at every entrance, which is unreasonable. Avoiding similar problems should be ensured by ut and e2e tests.

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml [hunk 2]
```diff
@@ -189,21 +189,39 @@
         select
         code
         from t_ds_workflow_definition
-        where project_code in
-            <foreach collection="projectCodes" index="index" item="i" open="(" separator="," close=")">
-                #{i}
-            </foreach>
+        <where>
```
Comment (ruanwenjun) [inline dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/WorkflowDefinitionMapper.xml]: Please revert this change, this kind of check should in `dao`, we need to avoid using complex logic in mapper.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProjectServiceImpl.java [hunk 3]
```diff
@@ -394,12 +394,15 @@ public Result queryProjectListPaging(User loginUser, Integer pageSize, Integer p
...
-                        ProjectWorkflowDefinitionCount::getCount));
+        List<Long> projectCodes = projectList.stream().map(Project::getCode).distinct().collect(Collectors.toList());
+        Map<Long, Integer> projectWorkflowDefinitionCountMap = Collections.emptyMap();
+        if (CollectionUtils.isNotEmpty(projectCodes)) {
+            projectWorkflowDefinitionCountMap = workflowDefinitionMapper
+                    .queryProjectWorkflowDefinitionCountByProjectCodes(projectCodes)
+                    .stream()
+                    .collect(Collectors.toMap(ProjectWorkflowDefinitionCount::getProjectCode,
+                            ProjectWorkflowDefinitionCount::getCount));
+        }
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProjectServiceImpl.java]: If the `projectList` is empty, then we can return early at line 389, then the logic would be clearer?

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Review by SbloodyS — APPROVED** [review]
+1
